core/src/main/java/com/datastrato/gravitino/catalog/FilesetEventDispatcher.java

```diff
@@ -0,0 +1,125 @@
+/*
+ *  Copyright 2024 Datastrato Pvt Ltd.
+ *  This software is licensed under the Apache License version 2.
+ */
+
+package com.datastrato.gravitino.catalog;
+
+import com.datastrato.gravitino.NameIdentifier;
+import com.datastrato.gravitino.Namespace;
+import com.datastrato.gravitino.exceptions.FilesetAlreadyExistsException;
+import com.datastrato.gravitino.exceptions.NoSuchFilesetException;
+import com.datastrato.gravitino.exceptions.NoSuchSchemaException;
+import com.datastrato.gravitino.file.Fileset;
+import com.datastrato.gravitino.file.FilesetChange;
+import com.datastrato.gravitino.listener.EventBus;
+import com.datastrato.gravitino.listener.api.event.AlterFilesetEvent;
+import com.datastrato.gravitino.listener.api.event.AlterFilesetFailureEvent;
+import com.datastrato.gravitino.listener.api.event.CreateFilesetEvent;
+import com.datastrato.gravitino.listener.api.event.CreateFilesetFailureEvent;
+import com.datastrato.gravitino.listener.api.event.DropFilesetEvent;
+import com.datastrato.gravitino.listener.api.event.DropFilesetFailureEvent;
+import com.datastrato.gravitino.listener.api.event.ListFilesetEvent;
+import com.datastrato.gravitino.listener.api.event.ListFilesetFailureEvent;
+import com.datastrato.gravitino.listener.api.event.LoadFilesetEvent;
+import com.datastrato.gravitino.listener.api.event.LoadFilesetFailureEvent;
+import com.datastrato.gravitino.listener.api.info.FilesetInfo;
+import com.datastrato.gravitino.utils.PrincipalUtils;
+import java.util.Map;
+
+/**
+ * {@code FilesetEventDispatcher} is a decorator for {@link FilesetDispatcher} that not only
+ * delegates fileset operations to the underlying catalog dispatcher but also dispatches
+ * corresponding events to an {@link EventBus} after each operation is completed. This allows for
+ * event-driven workflows or monitoring of fileset operations.
+ */
+public class FilesetEventDispatcher implements FilesetDispatcher {
+  private final EventBus eventBus;
+  private final FilesetDispatcher dispatcher;
+
+  public FilesetEventDispatcher(EventBus eventBus, FilesetDispatcher dispatcher) {
+    this.eventBus = eventBus;
+    this.dispatcher = dispatcher;
+  }
+
+  @Override
+  public NameIdentifier[] listFilesets(Namespace namespace) throws NoSuchSchemaException {
+    try {
+      NameIdentifier[] nameIdentifiers = dispatcher.listFilesets(namespace);
+      eventBus.dispatchEvent(new ListFilesetEvent(PrincipalUtils.getCurrentUserName(), namespace));
+      return nameIdentifiers;
+    } catch (Exception e) {
+      eventBus.dispatchEvent(
+          new ListFilesetFailureEvent(PrincipalUtils.getCurrentUserName(), namespace, e));
```

Comment:
I would suggest we can use an enum field to represent which type a FailureEvent instance means, the class of `FailureEvent` is too excessive.

Reply:
you mean use a general `FailureEvent` which an enum field to represent the type?

Reply:
Yeah.